### ISSUE TITLE
Regarding the training script

### ISSUE THREAD

**Issue by sayakpaul** [AdityaKane2001/regnety#8]
@AdityaKane2001 I looked into [the script](https://codeshare.io/gLLo40) you had shared in the call. Here are some points:

* Let's use two accuracies - top-1 accuracy and top-5 accuracy. Both can be logged using 
   ```
    metrics=[
            keras.metrics.CategoricalAccuracy(name="accuracy"),
            keras.metrics.TopKCategoricalAccuracy(5, name="top-5-accuracy")
        ]
   ```
   I also like the idea of using top-n error but I guess that would make it a bit more verbose. 
* To serialize the `.h5` checkpoints (only weights), I think we need to use `save_weights()` instead of `save()`. 
* `save_weights_only=True` would be better because it's common to first instantiate your model class and then loading the weights into it. 
* It'd also make sense to serialize the history object returned after model training to plot the model performance. 
* Since we'll be on TPUs`ModelCheckpoint` needs to operate with a GCS path rather than a local one. 
* Please don't use spaces in between the arguments and their values unless this is what's followed throughout the project. 
* Just to ensure things like bucket path, initial LR, optimizer, etc. -- will they be coming from a config file? 
* Are we not using any weight decay? If it's (or any other model regularization) not used in the original work, then it's fine.

**Comment by AdityaKane2001** [comment]
> I also like the idea of using top-n error but I guess that would make it a bit more verbose.

I just wanted to include top-1 since it is the metric of choice in the paper. Small question, but in top-1 case, error = 1 - accuracy, right?

> Please don't use spaces in between the arguments and their values unless this is what's followed throughout the project.

Sorry, will avoid that. Force of habit.

> Just to ensure things like bucket path, initial LR, optimizer, etc. -- will they be coming from a config file?

Yes. The variable `cfg` holds the required dataclass.

> Are we not using any weight decay?

I'll add that asap. Omitted optimizer definition for brevity.

I'll make the changes regarding model saving.

**Comment by sayakpaul** [comment]
> I just wanted to include top-1 since it is the metric of choice in the paper. Small question, but in top-1 case, error = 1 - accuracy, right?

Yes.

**Comment by AdityaKane2001** [comment]
@sayakpaul 

I'm a bit confused regarding weight decay and L2 regularization. The paper mentions usage of "weight decay" and in pycls, too the have added weight decay in the optimizer initialization. However, in TF we can do this using kernel_regularizer argument or we can use tfa.optimizers.SGDW.  The problem is that L2 and weight decay are being used interchangeably.
I believe L2 and weight decay are same for plain SGD, but are they same for SGD with momentum? If so, which implementation will be more accurate? 
I may be completely wrong here, because I am confused after going through many resources on this. 

(PS. By being the same I mean having same effect on the gradients)

**Comment by sayakpaul** [comment]
SGD with momentum is a different thing. See the relevant section of [this paper](https://arxiv.org/abs/2104.13963). 

We can use either I think -- SGDW (with a scheduler on both the LR and the decay factor) or by setting the `kernel_regularizer`. The latter is a bit more verbose in my opinion. So, we can start with the former one.

**Comment by AdityaKane2001** [comment]
>  SGDW (with a scheduler on both the LR and the decay factor)

Weight decay factor? It is constant according to pycls.
Do we want it to be 0 during warmup?

**Comment by sayakpaul** [comment]
Let's keep that as an option.

**Comment by AdityaKane2001** [comment]
If it's okay, can we close this one? I'll create a PR for training scripts soon.

**Comment by sayakpaul** [comment]
Yeah
